harness/determined/pytorch/_pytorch_context.py

```diff
@@ -477,12 +528,12 @@ def clip_grads(params):
         if self._should_communicate_and_update():
             # Communication needs to be synchronized so that is completed
             # before we apply gradient clipping and `step()`.
-            if self.hvd_config.use and not self._use_amp:
+            if self.hvd_config.use and not self._use_apex and not self._scaler:
                 optimizer.synchronize()  # type: ignore
```

Comment:
wait, does optimizer.synchronize ever get called when self._scaler is called()?

Reply:
Hmm... I actually don't remember now why optimizer.synchronize() is not called in the case of APEX and if that applies to AMP. We use skip_synchronize() down below when actually stepping the Horovod optimizer but that is done regardless of AMP.

Reply:
So it looks like we don't call it for APEX, because if APEX is enabled we actually make the call in backward(). This is apparently to unsure that we synchronize before unscaling.

However, the only reason I know of to unscale is right before you clip gradients, and we optionally do that for the user before calling their clip_grads function, or we document that they must call unscale at the beginning of that function. Both cases occur after this optimizer synchronize().

If the user would (for some reason unknown to me) need to unscale for some reason before calling step_optimizer, we would still need to do the same thing, actually. There's a comment that migrating to native PyTorch AMP addresses this but it doesn't explain further.

For now, I'm removing this change so that optimizer.synchronize() is called at least once for PyTorch. But I need to look into if there's another reason to call unscale() earlier.